File: ruby/hexadecimal_to_decimal_converter/hexadecimal_to_decimal_converter_4.rb
# Welcome, dear coder, to the mystical realm of hexadecimal to decimal conversion!
# This program is a wondrous tool designed to transform those cryptic hexadecimal numbers
# into their more familiar decimal counterparts. Prepare yourself for a journey through
# the enchanted lands of Ruby code, where variables and functions abound, and comments
# guide you like a trusty map.

# Function to convert a single hexadecimal digit to its decimal equivalent
def hex_digit_to_decimal(digit)
    # Behold, the magic of the hexadecimal digit transformation!
    case digit
    when '0'..'9'
        digit.ord - '0'.ord
    when 'a'..'f'
        digit.ord - 'a'.ord + 10
    when 'A'..'F'
        digit.ord - 'A'.ord + 10
    else
        # If the digit is not valid, we return the mystical -1
        -1
    end
end

# Function to convert a hexadecimal string to a decimal number
def hex_to_decimal(hex_string)
    # Let us embark on the grand conversion journey!
    decimal_value = 0
    power_of_sixteen = 1
    hex_string.reverse.each_char do |char|
        # Each character is a stepping stone in our quest
        decimal_digit = hex_digit_to_decimal(char)
        if decimal_digit == -1
            # If we encounter an invalid character, we must halt our journey
            return -1
        end
        decimal_value += decimal_digit * power_of_sixteen
        power_of_sixteen *= 16
    end
    decimal_value
end

# Function to display the result in a grand manner
def display_result(hex_string, decimal_value)
    # The grand reveal of the conversion result!
    puts "The hexadecimal number #{hex_string} converts to the decimal number #{decimal_value}."
end

# Main function to orchestrate the conversion symphony
def main
    # The main function is the conductor of our conversion symphony
    puts "Enter a hexadecimal number:"
    hex_string = gets.chomp
    decimal_value = hex_to_decimal(hex_string)
    if decimal_value == -1
        # If the conversion fails, we must inform the user of the unfortunate news
        puts "Invalid hexadecimal number entered."
    else
        # Otherwise, we display the glorious result
        display_result(hex_string, decimal_value)
    end
end

# The grand performance begins here
main

File: javascript/hexadecimal_to_decimal_converter/hexadecimal_to_decimal_converter_4.js
/**
 * This program is a magnificent creation designed to convert hexadecimal numbers
 * into their decimal equivalents. It is a splendid tool for those who wish to 
 * understand the inner workings of number systems and their conversions.
 * The program takes a hexadecimal string as input and returns the corresponding
 * decimal number. It is a marvel of modern programming, showcasing the elegance
 * and power of JavaScript.
 */

// Function to convert a single hexadecimal digit to its decimal equivalent
function hexDigitToDecimal(digit) {
    // The digit is converted to uppercase to handle both lowercase and uppercase inputs
    digit = digit.toUpperCase();
    // A variable to store the decimal value of the hexadecimal digit
    let decimalValue;
    // A switch statement to determine the decimal value based on the hexadecimal digit
    switch (digit) {
        case '0': decimalValue = 0; break;
        case '1': decimalValue = 1; break;
        case '2': decimalValue = 2; break;
        case '3': decimalValue = 3; break;
        case '4': decimalValue = 4; break;
        case '5': decimalValue = 5; break;
        case '6': decimalValue = 6; break;
        case '7': decimalValue = 7; break;
        case '8': decimalValue = 8; break;
        case '9': decimalValue = 9; break;
        case 'A': decimalValue = 10; break;
        case 'B': decimalValue = 11; break;
        case 'C': decimalValue = 12; break;
        case 'D': decimalValue = 13; break;
        case 'E': decimalValue = 14; break;
        case 'F': decimalValue = 15; break;
        default: decimalValue = -1; // Invalid hexadecimal digit
    }
    // The decimal value is returned to the caller
    return decimalValue;
}

// Function to convert a hexadecimal string to its decimal equivalent
function hexToDecimal(hexString) {
    // A variable to store the final decimal value
    let decimalValue = 0;
    // A variable to store the length of the hexadecimal string
    let